src/js/cli_engine.js

```diff
@@ -0,0 +1,458 @@
+    for (let i = 0, delay = 0; i < commandsArray.length; i++) {
+      let line = commandsArray[i].trim();
+      if (line.length === 0) continue;
+
+      if (delay) await new Promise((resolve) => setTimeout(resolve, delay));
+
+      if (this.#cliBufferContainsPartialCommand) {
+        line = this.#lineWithBuffer(line);
+      }
+
+      await new Promise((resolve) => this.sendLine(line, resolve));
+
+      this.#reportSendCommandsProgress((100.0 * i) / commandsArray.length);
+
+      if (line.toLowerCase().startsWith("profile")) {
+        delay = this.#profileSwitchDelayMs;
+      } else if (line.toLowerCase().startsWith("rateprofile")) {
+        delay = this.#rateProfileSwitchDelayMs;
+      } else {
+        delay = this.#lineDelayMs;
+      }
+    }
+    this.#reportSendCommandsProgress(100);
+  }
+
+  // removePromptHash removes the # from the prompt text, and is used to handle the output of the FC native autocomplete.
+  #removePromptHash(promptText) {
+    return promptText.replace(/^# /, "");
+  }
+
+  // The cliBuffer may have a partial command from a previous auto-complete request.
+  // #lineWithBuffer parses the input line and the cliBuffer, to return a line that when sent, will achieve the desired command.
+  // For example, if the cliBuffer contains "# dump all", and the user is trying to run a "diff all", the input for this function should be "diff all"
+  // and the line returned by this function would be `<number of backspaces required to reduce the cliBuffer to "# d">iff all`.
+  #lineWithBuffer(line) {
+    const buffer = this.#removePromptHash(this.#cliBuffer);
+    const bufferRegex = new RegExp(`^${buffer}`, "g");
+    if (line.match(bufferRegex)) {
+      return line.replace(bufferRegex, "");
+    }
+
+    let commonChars = 0;
+
+    for (let i = 0; i < buffer.length; i++) {
+      if (line[i] === buffer[i]) {
+        commonChars++;
+      } else {
+        break;
+      }
+    }
+
+    const noOfCharsToDelete = buffer.length - commonChars;
+
+    const backspace = String.fromCharCode(127);
+    return (
+      backspace.repeat(noOfCharsToDelete) +
+      line.substring(buffer.length - noOfCharsToDelete, line.length)
+    );
+  }
+
+  // writeToOutput is a function used to write text to the window wrapper -- while primarily used by the CliEngine, it is also used by CliAutoComplete.
+  writeToOutput(text) {
+    if (this.#cliAutoComplete && this.#cliAutoComplete.isBuilding()) {
+      this.#cliAutoComplete.builderParseLine(text);
+      return; // suppress output if the autocomplete is building state
+    }
+
+    this.#GUI.windowWrapper.append(text);
+    this.#GUI.window.scrollTop(this.#GUI.window.prop("scrollHeight"));
+  }
+
+  #writeLineToOutput(text) {
+    if (text.startsWith("###ERROR")) {
+      this.writeToOutput(`<span class="error_message">${text}</span><br>`);
+      this.#cliErrorsCount++;
+    } else {
+      this.writeToOutput(`${text}<br>`);
+    }
+    this.#responseCallback?.();
+  }
```

Comment:
This logic isn't correct. `builderParseLine()` needs to receive `text` without modification.